src/SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime/Design/SqlServerNodaTimeDesignTimeServices.cs

```diff
@@ -0,0 +1,23 @@
+﻿using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Design;
+using Microsoft.EntityFrameworkCore.Scaffolding;
+using Microsoft.EntityFrameworkCore.SqlServer.Storage;
+using Microsoft.EntityFrameworkCore.Storage;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace SimplerSoftware.EntityFrameworkCore.SqlServer.NodaTime.Design
+{
+    [UsedImplicitly]
+    public class SqlServerNodaTimeDesignTimeServices : IDesignTimeServices
+    {
+        public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
+            => serviceCollection
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, InstantTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, DurationTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, LocalDateTimeTypeMappingSourcePlugin>()
+                .AddSingleton<IRelationalTypeMappingSourcePlugin, OffsetDateTimeTypeMappingSourcePlugin>()
```

Comment:
I wonder if the default mappings for `datetime`, etc. will prevent these from being used...

Reply:
If so, you may need to override and replace the `SqlServerTypeMappingSource` service to avoid mapping them to DateTime and TimeSpan.

Reply:
Let me try again with the correct provider name. Thanks!

Reply:
Did not work! Maybe override SqlServerTypeMappingSource then?

Reply:
@bricelam - Thanks for the guidance!  I went ahead and committed your change regarding the provider name in the .targets file.  I can look into the `SqlServerTypeMappingSource` to see what needs to happen there.

Reply:
After looking at the `SqlServerTypeMappingSource` [here](https://github.com/dotnet/efcore/blob/34fa4029dc60168b5db16f7269f8f9645fedf592/src/EFCore.SqlServer/Storage/Internal/SqlServerTypeMappingSource.cs), I can see that the code will need to be refactored quite a bit.  When I wrote this I was trying to keep the code similar to what was already written in the EF Core codebase but it appears that during the 5.0 timeframe, the EF Core code was refactored quite a bit and consolidated all of the `IRelationalTypeMappingSourcePlugin`'s into a single class: `SqlServerTypeMappingSource` :)  The approach that I'm thinking of taking now is to create a new `NodaTimeSqlServerTypeMappingSource` class that inherits from the `SqlServerTypeMappingSource` class and consolidates all of the current NodaTime plugins into it.  Then in the `FindMapping` method, I will return a NodaTime relational mapping where appropriate, otherwise return the default `RelationTypeMapping` from the base implementation.  We'd have to then replace the default registration in the DI container with this new version. @bricelam - Does this sound like the appropriate way to go about this?

Reply:
> Does this sound like the appropriate way to go about this?

No. This seems like a unique case where you want an extension to override the *default* mapping for a store type (e.g. `datetime`) at design time. IMHO, the SqlServerTypeMappingSource should only be overridden at design time and not runtime. Overriding it at runtime could potentially create conflicts with other extensions or apps that also try to override it. In other words, keep the type mapping plugins for correct runtime behavior when the users adds NodaTime properties to their entities, but override SqlServerTypeMappingSource at design-time to force mapping to NodaTime types instead of BCL types when reverse engineering a model.

It all feels a bit brittle to me since the extensibility is really only designed to **add** type mappings, but doing it this way feels the least brittle.

Reply:
@roji Do you support mapping to NodaTime types during reverse engineer?

Reply:
> This seems like a unique case where you want an extension to override the default mapping for a store type (e.g. datetime) at design time.

Agreed :)  So instead of replacing the default provider across the board, I would only replace it during the reverse engineering process.  Does that just happen here in the `ConfigureDesignTimeServices` method?  If so then I *think* the method should look something like:
```csharp
public void ConfigureDesignTimeServices(IServiceCollection serviceCollection)
{
    // Search for the current default 'SqlServerTypeMappingSource'
    var mappingSourceDescriptor = serviceCollection.FirstOrDefault(x =>
        x.ServiceType == typeof(IRelationalTypeMappingSourcePlugin)
        && x.ImplementationType == typeof(SqlServerTypeMappingSource));

    // Determine where in the collection the NodaTime plugins should be inserted
    var insertionIndex = mappingSourceDescriptor != null
        ? serviceCollection.IndexOf(mappingSourceDescriptor)
        : serviceCollection.Count - 1;

    // These are either added before the current default plugin (if it exists) or the end of the collection
    // I'm assuming the EFCore internals iterate over all of the 'IRelationalTypeMappingSourcePlugin' registrations
    // and use the first valid 'RelationalTypeMapping' that is returned when searching for a valid type mapping.
    // Since these are in the collection *before* the default plugin, these should override the default where appropriate
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(InstantTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalDateTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(DurationTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(LocalDateTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(OffsetDateTimeTypeMappingSourcePlugin), ServiceLifetime.Singleton));
    serviceCollection.Insert(insertionIndex, new ServiceDescriptor(typeof(IRelationalTypeMappingSourcePlugin), typeof(SqlServerNodaTimeCodeGeneratorPlugin), ServiceLifetime.Singleton));
}
```

Does that *seem* correct?

Reply:
@StevenRasmussen it's important to distinguish between a provider's TypeMappingSource implementation, and a plugin-provided TypeMappingSourcePlugin. As @bricelam wrote above, you should not be trying to provide a TypeMappingSource, since that would have to replace SqlServerTypeMappingSource, which is undesirable; this is true for both runtime and design-time - there shouldn't be any reason to replace SqlServerTypeMappingSource in any situation. You should ideally just be providing a single SqlServerNodaTimeTypeMappingSourcePlugin (note the Plugin at the end).